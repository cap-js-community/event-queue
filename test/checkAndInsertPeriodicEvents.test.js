"use strict";

const path = require("path");

const cds = require("@sap/cds/lib");

const eventQueue = require("../src");
const { Logger: mockLogger } = require("./mocks/logger");
const { checkAndInsertPeriodicEvents } = require("../src/periodicEvents");
const { getConfigInstance } = require("../src/config");
const { selectEventQueueAndReturn } = require("./helper");
const project = __dirname + "/.."; // The project's root folder
cds.test(project);

describe("baseFunctionality", () => {
  let loggerMock, context, tx;
  beforeAll(async () => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date(1699873200000));
    const configFilePath = path.join(__dirname, "asset", "config.yml");
    await eventQueue.initialize({
      configFilePath,
      processEventsAfterPublish: false,
      registerAsEventProcessor: false,
      updatePeriodicEvents: false,
    });
    loggerMock = mockLogger();
    jest.spyOn(cds, "log").mockImplementation((layer) => {
      return mockLogger(layer);
    });
  });

  beforeEach(async () => {
    jest.clearAllMocks();
    context = new cds.EventContext({ user: "testUser", tenant: 123 });
    tx = cds.tx(context);
    await tx.run(DELETE.from("sap.eventqueue.Lock"));
    await tx.run(DELETE.from("sap.eventqueue.Event"));
  });

  afterEach(async () => {
    await tx.rollback();
  });

  it("basic insert all new events", async () => {
    await checkAndInsertPeriodicEvents(context);

    expect(loggerMock.callsLengths().error).toEqual(0);
    expect(loggerMock.calls().info).toMatchSnapshot();
    expect(await selectEventQueueAndReturn(tx)).toMatchSnapshot();
  });

  it("delta insert", async () => {
    await checkAndInsertPeriodicEvents(context);

    const fileContent = getConfigInstance().fileContent;
    fileContent.periodicEvents.push({
      ...getConfigInstance().fileContent.periodicEvents[0],
      subType: "DB2",
    });
    getConfigInstance().fileContent = fileContent;

    await checkAndInsertPeriodicEvents(context);

    fileContent.periodicEvents.splice(1, 1);
    getConfigInstance().fileContent = fileContent;
    expect(loggerMock.callsLengths().error).toEqual(0);
    expect(loggerMock.calls().info).toMatchSnapshot();
    expect(await selectEventQueueAndReturn(tx, 2)).toMatchSnapshot();
  });

  it("interval changed", async () => {
    await checkAndInsertPeriodicEvents(context);
    const eventConfig = getConfigInstance().periodicEvents[0];
    eventConfig.interval = 10;

    await tx.run(
      UPDATE.entity("sap.eventqueue.Event").set({
        startAfter: new Date(1699873200000 + 30 * 1000),
      })
    );
    await checkAndInsertPeriodicEvents(context);

    expect(loggerMock.callsLengths().error).toEqual(0);
    expect(loggerMock.calls().info).toMatchSnapshot();
    expect(await selectEventQueueAndReturn(tx, 1)).toMatchSnapshot();
  });
});
