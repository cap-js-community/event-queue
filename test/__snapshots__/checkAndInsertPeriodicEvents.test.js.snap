// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`baseFunctionality basic insert all new events 1`] = `
[
  [
    "inserting changed or new periodic events",
    {
      "events": [
        {
          "interval": 30,
          "subType": "DB",
          "type": "HealthCheck_PERIODIC",
        },
      ],
    },
  ],
]
`;

exports[`baseFunctionality basic insert all new events 2`] = `
[
  {
    "attempts": 0,
    "startAfter": "2023-11-13T11:00:00.000Z",
    "status": 0,
  },
]
`;

exports[`baseFunctionality delta insert 1`] = `
[
  [
    "inserting changed or new periodic events",
    {
      "events": [
        {
          "interval": 30,
          "subType": "DB",
          "type": "HealthCheck_PERIODIC",
        },
      ],
    },
  ],
  [
    "inserting changed or new periodic events",
    {
      "events": [
        {
          "interval": 30,
          "subType": "DB2",
          "type": "HealthCheck_PERIODIC",
        },
      ],
    },
  ],
]
`;

exports[`baseFunctionality delta insert 2`] = `
[
  {
    "attempts": 0,
    "startAfter": "2023-11-13T11:00:00.000Z",
    "status": 0,
  },
  {
    "attempts": 0,
    "startAfter": "2023-11-13T11:00:00.000Z",
    "status": 0,
  },
]
`;

exports[`baseFunctionality interval changed 1`] = `
[
  [
    "inserting changed or new periodic events",
    {
      "events": [
        {
          "interval": 30,
          "subType": "DB",
          "type": "HealthCheck_PERIODIC",
        },
      ],
    },
  ],
  [
    "deleting periodic events because they have changed",
    {
      "changedEvents": [
        {
          "subType": "DB",
          "type": "HealthCheck_PERIODIC",
        },
      ],
    },
  ],
  [
    "inserting changed or new periodic events",
    {
      "events": [
        {
          "interval": 10,
          "subType": "DB",
          "type": "HealthCheck_PERIODIC",
        },
      ],
    },
  ],
]
`;

exports[`baseFunctionality interval changed 2`] = `
[
  {
    "attempts": 0,
    "startAfter": "2023-11-13T11:00:00.000Z",
    "status": 0,
  },
]
`;
